jctools-core/src/main/java/org/jctools/queues/takestrategy/ParkTakeStrategy.java

```diff
@@ -0,0 +1,42 @@
+package org.jctools.queues.takestrategy;
+
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.LockSupport;
+
+public class ParkTakeStrategy<E> implements TakeStrategy<E>
+{
+
+    public volatile int storeFence = 0;
+
+    private AtomicReference<Thread> t               = new AtomicReference<Thread>(null);
+
+    @Override
+    public void signal()
+    {
+        // Make sure the offer is visible before unpark
+        storeFence = 1; // store barrier
```

Comment:
In the case of SPSC, the element is offered in the array with "putOrdered".
It's important that the unpark "happen after".
Otherwise there is a risk for the waiting thread to be unpark without seeing the element.
And so being reparked without processing the element.

Actually, you raised that problem to me a few months ago :)
I fixed it with a "putVolatile" in my "SPSCBlockingQueue" and I'm using a fence here to untouch the performance of the SPSC Queue when used in a non-blocking way.